In [5]:
# Смоделируем игру 

import numpy as np

trials = 10**6
successes = 0

good_luck = {777: 200, 999: 100, 555: 50, 333: 15, 111: 10}
x77 = [x for x in range(77, 1000, 100)]
xx7 = [x for x in range(7, 1000, 10)]
x00 = [x for x in range(0, 1000, 100)]
xx0 = [x for x in range(10, 1000, 10)]

win = 0
spent = 0

for i in range(trials):
    number = np.random.randint(0, 1000)
    if number in good_luck.keys():
        successes += 1
        win += good_luck[number]
        spent += 1
    elif number in x77:
        successes += 1
        win += 5
        spent += 1
    elif number in xx7:
        successes += 1
        win += 3
        spent += 1
    elif number in x00:
        successes += 1
        win += 2
        spent += 1
    elif number in xx0:
        successes += 1
        win += 1
        spent += 1
    else:
        spent += 1
    
p_monte_carlo = successes / trials
res = win - spent 
average_win = res / trials

print('Вероятность выигрыша из моделирования:', round(p_monte_carlo * 100, 2), '%')
print('Средний выигрыш/проигрыш:', round(average_win, 4), 'руб.')

Вероятность выигрыша из моделирования: 20.41 %
Средний выигрыш/проигрыш: -0.1977 руб.


In [3]:
# Посчитаем теоретическую вероятность выигрыша

good_luck_numbers = set(list(good_luck) + x77 + xx7 + x00 + xx0)
p = len(good_luck_numbers)/999
print('Теоретическая вероятность выигрыша:', round((p * 100), 2), '%')

Теоретическая вероятность выигрыша: 20.42 %


# Посчитаем теоретический средний выигрыш/проигрыш 

list_p = []

for x in good_luck.values():
    p_1 = x * 1 / 999
    list_p.append(p_1)

for x_ in good_luck_numbers:
    if x_ in x77:
        if x_ != 777:
            p_2 = 5 * 1 / 999
            list_p.append(p_2)
    elif x_ in xx7:
        if x_ != 777:
            if x_ != 77:
                p_2 = 3 * 1 / 999
                list_p.append(p_2)
    elif x_ in x00:
        p_2 = 2 * 1 / 999
        list_p.append(p_2)
    elif x_ in xx0:
        p_2 = 1 / 999
        list_p.append(p_2)
        
average_win = sum(list_p)
possible_game_res = round(average_win - (1 - p), 7)
print('Теоретически вероятный средний выигрыш/проигрыш:', possible_game_res, 'руб.')

### Вывод 

Вероятность выигрыша из моделирования и теоретических вычислений очень близки друг к другу: 20.41 % и 20.42 % соответственно.    

Что касается наиболее вероятного исхода игры, то модельная ситуация показала, что игрок остается в проигрыше почти на 20 копеек с каждой игры, что при большом количестве игр может обернуться большим проигрышем (в нашем случае Python "проиграл" 2 млн рублей). Исходя из того что в среднем ситуация является проигрышной для игрока, а проигрыш составляет пятую часть стоимости билета, от автомата лучше держаться подальше — игра не является выгодной, мягко говоря.